packages/grid/src/GridMetricCalculator.ts

```diff
@@ -47,6 +47,66 @@ export type GridMetricState = {
   isDraggingVerticalScrollBar: boolean;
 };
 
+/**
+ * Trim the provided map in place. Trims oldest inserted items down to the target size if the cache size is exceeded.
+ * Instead of trimming one item on every tick, we trim half the items so there isn't a cache clear on every new item.
+ * @param map The map to trim
+ * @param cacheSize The maximum number of elements to cache
+ * @param targetSize The number of elements to reduce the cache down to if `cacheSize` is exceeded
+ */
+export const trimMap = (
+  map: Map<unknown, unknown>,
+  cacheSize = GridMetricCalculator.CACHE_SIZE,
+  targetSize = Math.floor(cacheSize / 2)
+): void => {
+  if (map.size > cacheSize) {
+    const iter = map.keys();
+    while (map.size > targetSize) {
+      map.delete(iter.next().value);
+    }
+  }
+};
+
+/**
+ * Get the coordinates of floating items in one dimension.
+ * Can be used for getting the y coordinates of floating rows, or x coordinates of floating columns, calculated using the `sizeMap` passed in.
+ * @param startCount The number of floating items at the start (ie. `floatingTopRowCount` for rows, `floatingLeftColumnCount` for columns)
+ * @param endCount The number of floating items at the end (ie. `floatingBottomRowCount` for rows, `floatingRightColumnCount` for columns)
+ * @param totalCount Total number of items in this dimension (ie. `rowCount` for rows, `columnCount` for columns)
+ * @param max The max coordinate value (ie. `maxY` for rows, `maxX` for columns)
+ * @param sizeMap Map from index to size of item (ie. `rowHeightMap` for rows, `columnWidthMap` for columns)
+ */
+export const getFloatingCoordinates = (
+  startCount: number,
+  endCount: number,
+  totalCount: number,
+  max: number,
+  sizeMap: SizeMap
+): CoordinateMap => {
+  const coordinates = new Map();
+  let x = 0;
+  for (let i = 0; i < startCount && i < totalCount; i += 1) {
+    coordinates.set(i, x);
+    x += mapGet(sizeMap, i);
+  }
+
+  x = max;
+  for (let i = 0; i < endCount && totalCount - i - 1 >= 0; i += 1) {
+    x -= mapGet(sizeMap, totalCount - i - 1);
+    coordinates.set(totalCount - i - 1, x);
+  }
+  return coordinates;
+};
+
+function mapGet<K, V>(map: Map<K, V>, key: K): V {
+  const value = map.get(key);
+  if (value !== undefined) {
+    return value;
+  }
+
+  throw new Error(`Missing value for key ${key}`);
```

Comment:
I know you just changed this, but do we actually want to throw an error here? It would stop any future code in that function from running, so I guess it's just break earlier rather than later?

Reply:
I'll switch it to only throw in dev mode. The idea is that you use this function when the value _must_ be there

Reply:
Actually that just makes it complicated. This function should be used when you _know_ the value is there, we fail fast and hard if the value isn't there.
Calling map.get could return undefined, and rather than having to check that everywhere or provide a default value (which would mask the error), just fail hard